src/Evolution/Systems/Cce/LinearSolve.hpp

```diff
@@ -0,0 +1,322 @@
+    const ComplexDataVector& regular_integrand,
+    const ComplexDataVector& boundary, const ComplexDataVector& one_minus_y,
+    size_t l_max, size_t number_of_radial_points) noexcept;
+
+namespace detail {
+// needed because the standard transpose utility cannot create an arbitrary
+// ordering of blocks of data. This returns by pointer the configuration useful
+// for the linear solve step for H integration
+void transpose_to_reals_then_imags_radial_stripes(
+    gsl::not_null<DataVector*> result, const ComplexDataVector& input,
+    size_t number_of_radial_points, size_t number_of_angular_points) noexcept;
+}  // namespace detail
+
+// @{
+/*!
+ * \brief Computational structs for evaluating the hypersurface integrals during
+ * CCE evolution. These are compatible with use in `db::mutate_apply`.
+ *
+ * \details
+ * The integral evaluated and the corresponding inputs required depend on the
+ * CCE quantity being computed. In any of these, the only mutated tag is `Tag`,
+ * where the result of the integration is placed. The supported `Tag`s act in
+ * the following ways:
+ * - If the `Tag` is `Tags::BondiBeta` or `Tags::BondiU`, the integral to be
+ * evaluated is simply \f[ \partial_y f = A, \f] where \f$A\f$ is retrieved with
+ * `Tags::Integrand<Tag>`.
+ * - If the `Tag` is `Tags::BondiQ` or `Tags::BondiW`, the integral to be
+ * evaluated is \f[ (1 - y) \partial_y f + 2 f = A + (1 - y) B, \f] where
+ * \f$A\f$ is retrieved with `Tags::PoleOfIntegrand<Tag>` and \f$B\f$ is
+ * retrieved with `Tags::RegularIntegrand<Tag>`.
+ * - If `Tag` is `Tags::BondiH`, the integral to be evaluated is:
+ * \f[
+ * (1 - y) \partial_y f + L f + L^\prime \bar{f} = A + (1 - y) B,
+ * \f]
+ * where \f$A\f$ is retrieved with `Tags::PoleOfIntegrand<Tag>`, \f$B\f$ is
+ * retrieved with `Tags::RegularIntegrand<Tag>`, \f$L\f$ is retrieved with
+ * `Tags::LinearFactor<Tag>`, and \f$L^\prime\f$ is retrieved with
+ * `Tags::LinearFactorForConjugate<Tag>`. The presence of \f$L\f$ and
+ * \f$L^\prime\f$ ensure that the only current method we have for evaluating the
+ * \f$H\f$ hypersurface equation is a direct linear solve, rather than the
+ * spectral matrix multiplications which are available for the other integrals.
+ *
+ * In each case, the boundary value at the world tube for the integration is
+ * retrieved from `Tags::BoundaryValue<Tag>`.
+ *
+ * Additional type aliases `boundary_tags` and `integrand_tags` are provided for
+ * template processing of the required input tags necessary for these functions.
+ */
+template <typename Tag>
+struct RadialIntegrateBondi {
+  using boundary_tags = tmpl::list<Tags::BoundaryValue<Tag>>;
+  using integrand_tags = tmpl::list<Tags::Integrand<Tag>>;
+
+  using return_tags = tmpl::list<Tag>;
+  using argument_tags =
+      tmpl::append<integrand_tags, boundary_tags,
+                   tmpl::list<Spectral::Swsh::Tags::LMax,
+                              Spectral::Swsh::Tags::NumberOfRadialPoints>>;
+  static void apply(
```

Comment:
You can move the definitions of all of these `apply` functions into the cpp file and then generate instantiations of the `RadialIntegrateBondi` class for the supported tags.